code/game/jobs/job_exp.dm

```diff
@@ -0,0 +1,198 @@
+	if(!config.use_exp_tracking)
+		return "Tracking is disabled in the server configuration file."
+
+	var/list/play_records = params2list(prefs.exp)
+	if(!play_records.len)
+		return "[key] has no records."
+
+	var/return_text = "<UL>"
+
+	var/list/exp_data = list()
+	for(var/cat in exp_jobsmap)
+		if(text2num(play_records[cat]))
+			exp_data[cat] = text2num(play_records[cat])
+		else
+			exp_data[cat] = 0
+
+	for(var/dep in exp_data)
+		if(exp_data[dep] > 0)
+			if(dep == EXP_TYPE_EXEMPT)
+				return_text += "<LI>Exempt (all jobs auto-unlocked)</LI>"
+			else if(exp_data[EXP_TYPE_LIVING] > 0)
+				var/my_pc = num2text(round(exp_data[dep]/exp_data[EXP_TYPE_LIVING]*100))
+				return_text += "<LI>[dep] [get_exp_format(exp_data[dep])] ([my_pc]%)</LI>"
+			else
+				return_text += "<LI>[dep] [get_exp_format(exp_data[dep])] </LI>"
+	if(config.use_exp_restrictions_admin_bypass && check_rights(R_ADMIN, 0, mob))
+		return_text += "<LI>Admin (all jobs auto-unlocked)</LI>"
+	return_text += "</UL>"
+	var/list/jobs_locked = list()
+	var/list/jobs_unlocked = list()
+	for(var/datum/job/job in job_master.occupations)
+		if(job.exp_requirements && job.exp_type)
+			if(!job_is_xp_locked(job.title))
+				continue
+			else if(!job.available_in_playtime(mob.client))
+				jobs_unlocked += job.title
+			else
+				var/xp_req = job.exp_requirements
+				if(config.use_exp_restrictions_heads_hours && ((job.title in command_positions) || job.title == "AI"))
+					xp_req = config.use_exp_restrictions_heads_hours * 60
+				jobs_locked += "[job.title] [get_exp_format(text2num(play_records[job.exp_type]))] / [get_exp_format(xp_req)] [job.exp_type] EXP)"
+	if(jobs_unlocked.len)
+		return_text += "<BR><BR>Jobs Unlocked:<UL> "
+		for(var/text in jobs_unlocked)
+			return_text += "<LI>[text]</LI>"
+		return_text += "</UL>"
+	if(jobs_locked.len)
+		return_text += "<BR><BR>Jobs Not Unlocked:<UL>"
+		for(var/text in jobs_locked)
+			return_text += "<LI>[text]</LI>"
+		return_text += "</UL>"
+	return return_text
+
+
+/client/proc/get_exp_living()
+	var/list/play_records = params2list(prefs.exp)
+	var/exp_gen = text2num(play_records[EXP_TYPE_LIVING])
+	return get_exp_format(exp_gen)
+
+/proc/get_exp_format(var/expnum)
+	if(expnum > 60)
+		return num2text(round(expnum / 60)) + "h"
+	else if(expnum > 0)
+		return num2text(expnum) + "m"
+	else
+		return "0h"
+
+/proc/update_exp(var/mins, var/ann = 0)
+	if(!establish_db_connection())
+		return -1
+	spawn(0)
+		for(var/client/L in clients)
+			if(L.inactivity >= (10 MINUTES))
+				continue
+			spawn(0)
+				update_exp_client(L, mins, ann)
+			sleep(10)
+
+/proc/update_exp_client(var/client/C, var/minutes, var/announce_changes = 0)
+	if(!C ||!C.ckey)
+		return
+	var/DBQuery/exp_read = dbcon.NewQuery("SELECT exp FROM [format_table_name("player")] WHERE ckey='[C.ckey]'")
+	exp_read.Execute()
+	var/list/read_records = list()
+	var/hasread = 0
+	while(exp_read.NextRow())
+		read_records = params2list(exp_read.item[1])
+		hasread = 1
```

Comment:
This... isn't quite what I meant. You can check the return value of `Execute()`, and if it returns false, it means the query failed. Usually gets used like this:
```
if(!query.Execute())
	var/err = query.ErrorMsg()
	log_game("SQL ERROR during something. Error : \[[err]\]\n")
	message_admins("SQL ERROR during something. Error : \[[err]\]\n")
	return
```
Though... I guess this should also work, since `NextRow` certainly won't return any rows if the query errors out. Hm.